src/backend/context.rs

```diff
@@ -0,0 +1,706 @@
+        }
+    }
+
+    pub fn new_stream(&mut self,
+                      stream_name: *const i8,
+                      input_device: cubeb::DeviceId,
+                      input_stream_params: Option<cubeb::StreamParams>,
+                      output_device: cubeb::DeviceId,
+                      output_stream_params: Option<cubeb::StreamParams>,
+                      latency_frames: u32,
+                      data_callback: cubeb::DataCallback,
+                      state_callback: cubeb::StateCallback,
+                      user_ptr: *mut c_void) -> Result<Box<Stream>>
+    {
+        if self.error && self.pulse_context_init() != 0 {
+            return Err(cubeb::ERROR);
+        }
+
+        Stream::new(self,
+                    stream_name,
+                    input_device,
+                    input_stream_params,
+                    output_device,
+                    output_stream_params,
+                    latency_frames,
+                    data_callback,
+                    state_callback,
+                    user_ptr)
+    }
+
+    pub fn max_channel_count(&self) -> Result<u32>
+    {
+        unsafe {
+            pa_threaded_mainloop_lock(self.mainloop);
+            while self.default_sink_info.is_null() {
+                pa_threaded_mainloop_wait(self.mainloop);
+            }
+            pa_threaded_mainloop_unlock(self.mainloop);
+
+            Ok((*self.default_sink_info).channel_map.channels as u32)
+        }
+    }
+
+    pub fn preferred_sample_rate(&self) -> Result<u32>
+    {
+        pa_threaded_mainloop_lock(self.mainloop);
+        while self.default_sink_info.is_null() {
+            pa_threaded_mainloop_wait(self.mainloop);
+        }
+        pa_threaded_mainloop_unlock(self.mainloop);
+
+        unsafe {
+            Ok((*self.default_sink_info).sample_spec.rate)
+        }
+    }
+
+    pub fn min_latency(&self, params: &cubeb::StreamParams) -> Result<u32>
+    {
+        // According to PulseAudio developers, this is a safe minimum.
+        Ok(25 * params.rate / 1000)
+    }
+
+    pub fn preferred_channel_layout(&self) -> Result<cubeb::ChannelLayout>
+    {
+        pa_threaded_mainloop_lock(self.mainloop);
+        while self.default_sink_info.is_null() {
+            pa_threaded_mainloop_wait(self.mainloop);
+        }
+        pa_threaded_mainloop_unlock(self.mainloop);
+
+        unsafe {
+            Ok(channel_map_to_layout(&(*self.default_sink_info).channel_map))
+        }
+    }
+
+    pub fn enumerate_devices(&self, devtype: cubeb::DeviceType) -> Result<*mut cubeb::DeviceCollection>
+    {
+        let mut user_data: PulseDevListData = Default::default();
+        user_data.context = self as *const _ as *mut _;
+
+        pa_threaded_mainloop_lock(self.mainloop);
+
+        let o = pa_context_get_server_info(self.context,
+                                           Some(pulse_server_info_cb),
+                                           &mut user_data as *mut _ as *mut _);
```

Comment:
Yeah, unfortunately they do.

`&mut user_data as *mut _` results in type `*mut PulseDevListData`, the second `as *mut _` results in type `*mut c_void`. You can't jump from `&mut t` to `*mut c_void` in one step in rust, otherwise rustc will report error:

casting `&mut backend::context::PulseDevListData` as `*mut std::os::raw::c_void` is invalid